[LSR] Recognize vscale-relative immediates

huntergr-arm opened the pull request:
Final part of the vscale-aware LSR work, see https://discourse.llvm.org/t/rfc-vscale-aware-loopstrengthreduce/77131

It's a bit messy right now, I mainly just want to know if there's any objections to the current work before I finish it up.

huntergr-arm commented:
Rebased on top of the getScalingFactorCost patch.

SamTebbs33 reviewed (commented):
I like it so far. It seems like most of the changes are there to move from integers to Immediates.

huntergr-arm committed:
Test cases for vscale immediates

huntergr-arm committed:
Scalable work

huntergr-arm committed:
Convert LSR to use possibly-scalable Immediate type

huntergr-arm commented:
Separated the conversion to use Immediate from the changes to support vscale immediates into different commits.

huntergr-arm commented:
There's still several places where `getFixedValue()` is called where we might see a scalable immediate. I'll try writing some tests to see if I can reach them and fix them.

huntergr-arm commented:
Added some new test cases and converted the checks to look at llc output, since it's easier to see what the impact is. I think the cost comparison changes will also be required to get the best results here, as there's still some problems with the many_mulvl1_addressing output.

davemgreen commented:
Hi - I have been looking at something that is related but mostly orthogonal to this in the generation of chains for vscale addressing modes recently. I've put the patch up at #94453. I don't think it changes alot of the same code as this patch, which is altering a lot more. That probably turned out to mostly be about how chains are generated. It does make use of the existing isLegalAddressingMode changes which have made it a lot easier since the last time I looked into it.

My high level comment for this patch, after looking into the chains, is the same as it has been elsewhere - that it feels like it is useful to at least conceptually be able to model `Base + Imm + vscale * ScalableImm` with both an Offset and a ScalableOffset treated independently. Even if it is unlikely that they are supported by the assembly, being able to represent them so that it can reason about things like `(Base+Imm) + vscale*ScalableImm` vs `(Base+vscale*ScalableImm) + imm` can be useful. I obviously could be wrong though, as I haven't seen the alternative version and am not sure how difficult it would be in practice.

huntergr-arm commented:
Rebased on top of the isLSRCostLess changes I merged.

huntergr-arm commented:
> Hi - I have been looking at something that is related but mostly orthogonal to this in the generation of chains for vscale addressing modes recently. I've put the patch up at #94453. I don't think it changes alot of the same code as this patch, which is altering a lot more. That probably turned out to mostly be about how chains are generated. It does make use of the existing isLegalAddressingMode changes which have made it a lot easier since the last time I looked into it.
> 
> My high level comment for this patch, after looking into the chains, is the same as it has been elsewhere - that it feels like it is useful to at least conceptually be able to model `Base + Imm + vscale * ScalableImm` with both an Offset and a ScalableOffset treated independently. Even if it is unlikely that they are supported by the assembly, being able to represent them so that it can reason about things like `(Base+Imm) + vscale*ScalableImm` vs `(Base+vscale*ScalableImm) + imm` can be useful. I obviously could be wrong though, as I haven't seen the alternative version and am not sure how difficult it would be in practice.

I'm assuming by independently representing fixed and scalable immediates you mean we would have (potentially) two separate instances of `Immediate` (or whatever else we end up using)? I suppose there could be a use in terms of finding a legal add immediate followed up with a legal folded addressing mode, but I think that's outside the scope of this patch (which is just aimed at getting vscale-relative immediates off the ground).

SamTebbs33 reviewed (approved):
Looks good to me.

paulwalker-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -4004,27 +4160,29 @@ void LSRInstance::GenerateICmpZeroScales(LSRUse &LU, unsigned LUIdx,
+    if (Base.BaseOffset.isMin() && Factor == -1)
       continue;
-    int64_t NewBaseOffset = (uint64_t)Base.BaseOffset * Factor;
+    Immediate NewBaseOffset =
+        Immediate::getFixed((uint64_t)Base.BaseOffset.getFixedValue() * Factor);
```
Similar to above, `FixedOrScalableQuantity` supports `operator*` with a scalar multiplier.

paulwalker-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -4004,27 +4160,29 @@ void LSRInstance::GenerateICmpZeroScales(LSRUse &LU, unsigned LUIdx,
       continue;
-    Offset = (uint64_t)Offset * Factor;
-    if (Offset / Factor != LU.MinOffset)
+    Offset = Immediate::getFixed((uint64_t)Offset.getFixedValue() * Factor);
+    if (Offset.getFixedValue() / Factor != LU.MinOffset.getFixedValue())
```
Is this path fixed-offset specific or it is the case we've just not hit it for scalable-offsets yet?

huntergr-arm replied:
Fixed specific, it's generating terms for ICmpZero uses (for which we won't have any legal scalable offsets)

paulwalker-arm reviewed (commented):
I'm not going to bang the "we should be using StackOffset everywhere" drum this time but there are certainly places within this patch where it would make things safer.  

The hardest part of reviewing the patch is all the `getFixedValue()` calls where it's not clear why it safe to assume a fixed offset.  There are part of the code where I think we should be able to use the operator overloading better to make the code agnostic to being fixed/scalable.  Even for parts that are specific to one of those I think the code would be cleaner if it didn't look special.

Do you think it's possible to incorporate a scalable safe type without necessarily adding the scalable safe code paths?  I'm not asking for a separate PR but the genuine improvements that would be easier to spot if they were in a separate commit separate from the refactoring.

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1401,24 +1483,29 @@ void Cost::RateFormula(const Formula &F,
   for (const LSRFixup &Fixup : LU.Fixups) {
-    int64_t O = Fixup.Offset;
-    int64_t Offset = (uint64_t)O + F.BaseOffset;
+    // FIXME: We probably want to noticeably increase the cost if the
+    // two offsets differ in scalability?
+    bool Scalable = Fixup.Offset.isScalable() || F.BaseOffset.isScalable();
+    int64_t O = Fixup.Offset.getKnownMinValue();
+    Immediate Offset = Immediate::get(
+        (uint64_t)(O) + F.BaseOffset.getKnownMinValue(), Scalable);
```
I think it's more that a wrapper for the casted arithmetic is needed. Right now we would need to cast and add the scalable and fixed parts of StackOffset independently.

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -3931,21 +4082,26 @@ void LSRInstance::GenerateConstantOffsetsImpl(
     return;
   Formula F = Base;
-  F.BaseOffset = (uint64_t)F.BaseOffset + Imm;
+  F.BaseOffset = Immediate::get((uint64_t)F.BaseOffset.getKnownMinValue() +
+                                    Imm.getKnownMinValue(),
+                                Imm.isScalable());
```
It's the cast to unsigned that breaks the operators. (The internal quantity type is int64_t)

paulwalker-arm replied:
Assuming the cast is important, what about implementing something like `Immediate::AddToUnsigned(Immediate&)`?  I'd sooner something explicit that works for both fixed and scalable rather than continually pulling apart the `Immediate` type.

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -3931,21 +4082,26 @@ void LSRInstance::GenerateConstantOffsetsImpl(
-          Offset -= Step;
-          GenerateOffset(G, Offset);
+        for (Immediate Offset : Worklist) {
+          if (!Offset.isScalable()) {
```
It's trying to subtract an APInt derived from a fixed SCEVAddRecExpr step. We may be able to check for a scalable step to make use of this, but I'd want to come up with an appropriate test first.

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1308,9 +1390,9 @@ void Cost::RateRegister(const Formula &F, const SCEV *Reg,
       // If the step size matches the base offset, we could use pre-indexed
       // addressing.
-      if (AMK == TTI::AMK_PreIndexed) {
+      if (AMK == TTI::AMK_PreIndexed && !F.BaseOffset.isScalable()) {
```
There's no 'isFixed()' method on FixedOrScalableQuantity, so I just used the inverse of isScalable(). They're logically equivalent, though there's semantic reinforcement from separating them.

But I do wonder if we want an uninitialized state for scalable-ness, and allow that to be freely used with either fixed or scalable offsets. The result would then be 'locked in' to one or the other.

paulwalker-arm replied:
I'm suggesting to add the new function because they're not logically equivalent. It's just the "zero" support is incomplete and my worry is that once that's fixed, code like this will silently break.

huntergr-arm replied:
Immediate::isZero() is the same function you've referenced -- it ignores the Scalable flag and just returns Quantity. isScalable() returns Scalable, and a new isFixed() would return '!Scalable'.

Do we want isScalable() to mean 'Scalable || Quantity == isZero()' and change the other FixedOrScalableQuantity types too?

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -2590,18 +2720,25 @@ bool LSRInstance::reconcileNewOffset(LSRUse &LU, int64_t NewOffset,
 
   // Conservatively assume HasBaseReg is true for now.
-  if (NewOffset < LU.MinOffset) {
+  if (Immediate::isKnownLT(NewOffset, LU.MinOffset)) {
```
Maybe we need a comparison function which returns a std::optional around the result? So we can distinguish between 'no, it's not less than' and 'I cannot tell you if it's less than'.

huntergr-arm commented:
> I'm not going to bang the "we should be using StackOffset everywhere" drum this time but there are certainly places within this patch where it would make things safer.
> 
> The hardest part of reviewing the patch is all the `getFixedValue()` calls where it's not clear why it safe to assume a fixed offset. There are part of the code where I think we should be able to use the operator overloading better to make the code agnostic to being fixed/scalable. Even for parts that are specific to one of those I think the code would be cleaner if it didn't look special.
> 
> Do you think it's possible to incorporate a scalable safe type without necessarily adding the scalable safe code paths? I'm not asking for a separate PR but the genuine improvements that would be easier to spot if they were in a separate commit separate from the refactoring.

I'm not strictly opposed to using StackOffset, and it may simplify some of the other parts of the code. I used a mutually-exclusive struct here because there are places where it assumes it can make a combined offset between the current known min and max that must be legal, and asserts as much. If we cannot establish ordering between scalable and fixed offsets, then the whole min/max thing breaks.

So we could maybe use StackOffset, but basically use them in the same way Immediate is now, and track scalable and fixed min/max separately.

huntergr-arm commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -3887,15 +4027,26 @@ void LSRInstance::GenerateSymbolicOffsets(LSRUse &LU, unsigned LUIdx,
+  auto GenerateOffset = [&](const SCEV *G, Immediate Offset) {
     Formula F = Base;
-    F.BaseOffset = (uint64_t)Base.BaseOffset - Offset;
+    if (Base.BaseOffset.isScalable() != Offset.isScalable() &&
+        Base.BaseOffset.isNonZero() && Offset.isNonZero())
```
Yeah, I treat them that way as well. But this is checking for scalability being equal _if_ both are nonzero. I guess an 'isCompatible()' helper might be better (subject to bikeshedding over the name).

huntergr-arm commented:
> Do you think it's possible to incorporate a scalable safe type without necessarily adding the scalable safe code paths? I'm not asking for a separate PR but the genuine improvements that would be easier to spot if they were in a separate commit separate from the refactoring.

It is in a separate commit ('Convert LSR to use possibly-scalable Immediate type', followed by 'Scalable work')

huntergr-arm committed:
Add isFixed, getZero, getScalableFixedMin/Max convenience methods

huntergr-arm commented:
Rebased, added some convenience functions.

I didn't switch to StackOffset yet. I think handling cases with mixed offsets requires more thought, and possibly redoing how we handle initial Use generation. When we do switch to it though, we can move the SCEV generation code to ScalarEvolution since that's an existing class available across the codebase.

SamTebbs33 reviewed (approved):
Still looks good to me with your latest changes.
